ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala

```diff
@@ -133,73 +148,74 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
-    discard(
-      SQL"insert into #$TableName(#$LedgerIdColumnName) values($ledgerId)".execute()(connection)
-    )
-
-  def participantId(connection: Connection): Option[ParticipantId] =
-    SQL"select #$ParticipantIdColumnName from #$TableName".as(ParticipantIdParser.single)(
-      connection
-    )
-
-  def updateParticipantId(participantId: String)(connection: Connection): Unit =
-    discard(
-      SQL"update #$TableName set #$ParticipantIdColumnName = $participantId".execute()(
-        connection
-      )
-    )
-
-  def ledgerEndOffset(connection: Connection): Offset =
-    SelectLedgerEnd.as(LedgerEndOrBeforeBeginParser.single)(connection)
-
-  // TODO mutable contract state cache - use only one getLedgerEnd method
-  def ledgerEndOffsetAndSequentialId(connection: Connection): (Offset, Long) =
-    SelectLedgerEndOffsetAndSequentialId.as(LedgerEndOffsetAndSequentialIdParser.single)(connection)
+  private val LedgerIdentityParser: RowParser[OptionalIdentityParams] =
+    LedgerIdParser ~ ParticipantIdParser map { case ledgerId ~ participantId =>
+      OptionalIdentityParams(Some(ledgerId), participantId)
+    }
 
-  def initialLedgerEnd(connection: Connection): Option[Offset] =
-    SelectLedgerEnd.as(LedgerEndParser.single)(connection)
+  override def initializeParameters(
+      params: StorageBackend.IdentityParams
+  )(connection: Connection)(implicit loggingContext: LoggingContext): Unit = {
+    // Note: this method is the only one that inserts a row into the parameters table
+    val previous = ledgerIdentity(connection)
+    val ledgerId = params.ledgerId
+    val participantId = params.participantId
+    previous match {
+      case StorageBackend.OptionalIdentityParams(None, _) =>
+        // ledgerId is not null, this is the case where the the parameters table is empty
+        logger.info(
+          s"Initializing new database for ledgerId '${params.ledgerId}' and participantId '${params.participantId}'"
+        )
+        discard(
+          SQL"insert into #$TableName(#$LedgerIdColumnName, #$ParticipantIdColumnName) values(${ledgerId.unwrap}, ${participantId.unwrap: String})"
+            .execute()(connection)
+        )
+      case StorageBackend.OptionalIdentityParams(Some(`ledgerId`), None) =>
+        logger.info(
+          s"Found existing database for ledgerId '${params.ledgerId}', initializing participantId '${params.participantId}'"
+        )
+        discard(
+          SQL"update #$TableName set #$ParticipantIdColumnName=${participantId.unwrap: String}"
+            .execute()(connection)
+        )
+      case StorageBackend.OptionalIdentityParams(Some(`ledgerId`), Some(`participantId`)) =>
+        logger.info(
+          s"Found existing database for ledgerId '${params.ledgerId}' and participantId '${params.participantId}'"
+        )
+      case StorageBackend.OptionalIdentityParams(_, Some(existing))
+          if existing != params.participantId =>
```

Comment:
I think if you would move the line204 case up, then you would not need to check for equality here

Reply:
Participant id mismatch has a higher priority than ledger id mismatch (in cases where both mismatch).

Reply:
are you sure? looking at the original indexer code I had another impression :thinking: 
also: I am not sure it matters too much, either way we are crashing, no?
(also: it seem to me both is possible as well, which would be quite hard to express with the current exceptions...I am not arguing to change that, just an observation)

Reply:
Actually, you are right, I have moved the lines around. Thanks for catching this!